src/main/java/it/smartcommunitylab/aac/saml/provider/SamlAuthenticationProvider.java

```diff
@@ -108,7 +112,35 @@ public SamlAuthenticationProvider(
         //                    params.put(SAML2AssertionValidationParameters.CLOCK_SKEW, Duration.ofMinutes(5).toMillis());
         //                    return new ValidationContext(params);
         //                }));
-
+        Converter<OpenSaml4AuthenticationProvider.ResponseToken, Saml2Authentication> defaultAssertionConverter =
+            OpenSaml4AuthenticationProvider.createDefaultResponseAuthenticationConverter();
+        openSamlProvider.setResponseAuthenticationConverter(responseToken -> {
+            // use default converter to pre-evaluate authentication, containing attributes and indexes
+            // then, extract AuthnContextClassRef from Assertion and include it as a custom attribute
+            Response response = responseToken.getResponse();
+            Saml2AuthenticationToken token = responseToken.getToken();
+            Saml2Authentication auth = defaultAssertionConverter.convert(responseToken);
+            Map<String, List<Object>> attributes = new HashMap<>(
+                ((Saml2AuthenticatedPrincipal) auth.getPrincipal()).getAttributes()
+            );
+            List<Object> acrValues = extractAcrAssertion(response);
+            if ((acrValues != null) && (!attributes.containsKey(ACR_ATTRIBUTE))) {
```

Comment:
set the value in attributes regardless of the previous values: if you read the assertion it doesn't matter that there is already a value for that in the result, you overwrite it.